File: sdk-kotlin/src/test/kotlin/com/hobostays/rtdb/core/FramesTest.kt
package com.hobostays.rtdb.core

import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNull
import kotlin.test.assertTrue
import org.junit.jupiter.api.Test

/** §Transport tolerance rules and the §1–§10 frame shapes, ported from WP1's `frames.test.ts`. */
class FramesTest {

    @Test
    fun `every client frame encodes as one object with a type field`() {
        assertEquals(
            """{"type":"hello","token":"t","proto":1,"sdk":"kotlin/0.1.0"}""",
            Hello(token = "t", sdk = "kotlin/0.1.0").encode(),
        )
        assertEquals("""{"type":"unlisten","subId":7}""", Unlisten(subId = 7).encode())
        assertEquals("""{"type":"ping","t":42}""", Ping(t = 42).encode())
        assertEquals(
            """{"type":"cas","writeId":"w","path":"p/score","expectedRev":184224,"value":51}""",
            Cas(writeId = "w", path = "p/score", expectedRev = 184224, value = JsonPrimitive(51)).encode(),
        )
    }

    @Test
    fun `an absent lastRev is omitted, not sent as null (§3 no-or-zero means snapshot)`() {
        assertEquals("""{"type":"listen","subId":1,"path":"room"}""", Listen(subId = 1, path = "room").encode())
        assertEquals(
            """{"type":"listen","subId":1,"path":"room","lastRev":184190}""",
            Listen(subId = 1, path = "room", lastRev = 184190).encode(),
        )
    }

    @Test
    fun `a put carries null verbatim — it is the wire form of removeValue (§4)`() {
        assertEquals(
            """{"type":"put","writeId":"w","path":"p","value":null}""",
            Put(writeId = "w", path = "p", value = JsonNull).encode(),
        )
    }

    @Test
    fun `unknown fields are ignored (§Transport)`() {
        val frame = parseServerFrame(
            """{"type":"helloAck","rev":7,"epoch":3,"region":"r","session":"s_1","futureField":{"x":1}}""",
        )
        assertEquals(HelloAck(rev = 7, epoch = 3, region = "r", session = "s_1"), frame)
    }

    @Test
    fun `unknown frame types are ignored, never errors (§Transport)`() {
        assertNull(parseServerFrame("""{"type":"getAck","reqId":3,"value":1,"rev":9}"""), "§11 extension frame")
        assertNull(parseServerFrame("""{"type":"somethingFromV2"}"""))
        assertNull(parseServerFrame("not json at all"))
        assertNull(parseServerFrame("""{"no":"type"}"""))
        assertNull(parseServerFrame("""{"type":"ack","writeId":"w"}"""), "a known type missing a field")
    }

    @Test
    fun `an unknown frame inside a batch does not take its batch-mates with it (§3)`() {
        val batch = parseServerFrame(
            """{"type":"batch","frames":[
                {"type":"delta","rev":1,"path":"p","op":"put","value":1},
                {"type":"fromTheFuture","x":1},
                {"type":"ack","writeId":"w","rev":2}
            ]}""",
        )
        assertIs<Batch>(batch)
        val inner = batch.frames.mapNotNull { parseServerFrame(it) }
        assertEquals(listOf(Delta(rev = 1, path = "p", op = "put", value = JsonPrimitive(1)), Ack("w", 2)), inner)
    }

    @Test
    fun `a delta keeps an unknown op instead of failing to parse (§3 re-listen rule)`() {
        val delta = parseServerFrame("""{"type":"delta","rev":9,"path":"p","op":"incr","value":5}""")
        assertIs<Delta>(delta)
        assertEquals("incr", delta.op, "the sub goes stale and re-listens; the frame itself still parses")
    }

    @Test
    fun `err parses at every scope (§2, §3, §4)`() {
        assertEquals(Err(code = "AUTH", msg = "token expired"), parseServerFrame("""{"type":"err","code":"AUTH","msg":"token expired"}"""))
        assertEquals(Err(code = "RULES", msg = "read denied", subId = 7), parseServerFrame("""{"type":"err","code":"RULES","msg":"read denied","subId":7}"""))
        assertEquals(Err(code = "RATE", msg = "", writeId = "w"), parseServerFrame("""{"type":"err","code":"RATE","writeId":"w"}"""))
    }

    @Test
    fun `server frames round-trip the values the mirror will need (§3, §4)`() {
        val snapshot = parseServerFrame("""{"type":"snapshot","subId":7,"path":"p","value":{"name":"Ravi","score":42},"rev":184223}""")
        assertIs<Snapshot>(snapshot)
        assertEquals(buildJsonObject { put("name", JsonPrimitive("Ravi")); put("score", JsonPrimitive(42)) }, snapshot.value)
        assertEquals(184223, snapshot.rev)

        assertEquals(Resync(subId = 7), parseServerFrame("""{"type":"resync","subId":7}"""))
        assertEquals(Pong(t = 1756280000000), parseServerFrame("""{"type":"pong","t":1756280000000}"""))
        val casFail = parseServerFrame("""{"type":"casFail","writeId":"w","path":"p","value":50,"rev":184224}""")
        assertIs<CasFail>(casFail)
        assertEquals(JsonPrimitive(50), casFail.value)
    }

    @Test
    fun `a merge value is an object of possibly deep keys (§4)`() {
        val merge = Merge(
            writeId = "w",
            path = "p",
            value = buildJsonObject {
                put("score", JsonPrimitive(50))
                put("stats/wins", JsonPrimitive(3))
                put("tag", JsonNull)
            },
        )
        assertTrue(merge.encode().contains(""""stats/wins":3"""))
        assertTrue(merge.encode().contains(""""tag":null"""), "a null child deletes; it must survive encoding")
    }
}

File: test/unit/frames.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { DEFAULT_LIMITS } from '../../src/protocol/limits.ts';
import { parseClientFrame, type ParseResult } from '../../src/protocol/validate.ts';

const W = '0d0e1f2a-3b4c-4d5e-8f90-a1b2c3d4e5f6';
const parse = (v: unknown): ParseResult =>
  parseClientFrame(typeof v === 'string' ? v : JSON.stringify(v), DEFAULT_LIMITS);

const frame = (v: unknown) => {
  const r = parse(v);
  assert.equal(r.kind, 'frame', `expected a frame, got ${JSON.stringify(r)}`);
  return (r as Extract<ParseResult, { kind: 'frame' }>).frame;
};
const reject = (v: unknown) => {
  const r = parse(v);
  assert.equal(r.kind, 'reject', `expected a reject, got ${JSON.stringify(r)}`);
  return r as Extract<ParseResult, { kind: 'reject' }>;
};

// ---- §Transport tolerance rules -------------------------------------------------

test('unknown frame types are ignored, never errors', () => {
  for (const type of ['reauth', 'get', 'incr', 'somethingFromV3']) {
    assert.equal(parse({ type, whatever: 1 }).kind, 'ignore');
  }
});

test('unknown fields are ignored and stripped', () => {
  assert.deepEqual(frame({ type: 'ping', t: 7, futureField: { a: 1 } }), { type: 'ping', t: 7 });
  assert.deepEqual(frame({ type: 'hello', proto: 1, token: 'tk', sdk: 'kotlin/1.0.0', region: 'x' }), {
    type: 'hello',
    proto: 1,
    token: 'tk',
    sdk: 'kotlin/1.0.0',
  });
});

test('non-frames are BADFRAME', () => {
  for (const raw of ['{', 'null', '3', '"str"', '[{"type":"ping","t":1}]']) {
    assert.equal(reject(raw).code, 'BADFRAME');
  }
  assert.equal(reject({ noType: true }).code, 'BADFRAME');
  assert.equal(reject({ type: 9 }).code, 'BADFRAME');
});

// ---- §2 hello --------------------------------------------------------------------

test('hello requires proto and token; sdk is optional', () => {
  assert.deepEqual(frame({ type: 'hello', proto: 1, token: 'tk' }), { type: 'hello', proto: 1, token: 'tk' });
  reject({ type: 'hello', token: 'tk' });
  reject({ type: 'hello', proto: 1 });
  reject({ type: 'hello', proto: 0, token: 'tk' });
  reject({ type: 'hello', proto: 1, token: 5 });
});

// ---- §3 listen / unlisten --------------------------------------------------------

test('listen validates subId, path and lastRev', () => {
  assert.deepEqual(frame({ type: 'listen', subId: 7, path: 'a/b', lastRev: 184190 }), {
    type: 'listen',
    subId: 7,
    path: 'a/b',
    lastRev: 184190,
  });
  // §3: no/zero lastRev both mean "send a snapshot" — 0 normalizes to absent.
  assert.deepEqual(frame({ type: 'listen', subId: 7, path: '' }), { type: 'listen', subId: 7, path: '' });
  assert.deepEqual(frame({ type: 'listen', subId: 7, path: '', lastRev: 0 }), {
    type: 'listen',
    subId: 7,
    path: '',
  });
  reject({ type: 'listen', subId: 0, path: 'a' });
  reject({ type: 'listen', subId: 1.5, path: 'a' });
  reject({ type: 'listen', subId: 7, path: 'a', lastRev: -1 });
});

test('a bad listen path is BADPATH scoped to the subId', () => {
  const r = reject({ type: 'listen', subId: 7, path: 'a//b' });
  assert.equal(r.code, 'BADPATH');
  assert.equal(r.subId, 7);
});

test('unlisten needs only a subId', () => {
  assert.deepEqual(frame({ type: 'unlisten', subId: 7 }), { type: 'unlisten', subId: 7 });
  reject({ type: 'unlisten' });
});

// ---- §4 writes -------------------------------------------------------------------

test('put carries any JSON value, and null means delete', () => {
  assert.deepEqual(frame({ type: 'put', writeId: W, path: 'a', value: null }), {
    type: 'put',
    writeId: W,
    path: 'a',
    value: null,
  });
  assert.deepEqual(frame({ type: 'put', writeId: W, path: '', value: { n: 'Ravi' } }).type, 'put');
  // absent value is not the same as null
  assert.equal(reject({ type: 'put', writeId: W, path: 'a' }).code, 'BADFRAME');
});

test('writeId must be a UUID, and write rejects are scoped to it', () => {
  assert.equal(reject({ type: 'put', writeId: 'nope', path: 'a', value: 1 }).code, 'BADFRAME');
  assert.equal(reject({ type: 'put', writeId: W, path: 'a/', value: 1 }).writeId, W);
});

test('merge value must be an object; deep keys are validated as paths', () => {
  assert.deepEqual(frame({ type: 'merge', writeId: W, path: 'a', value: { 'stats/wins': 3, tag: null } }), {
    type: 'merge',
    writeId: W,
    path: 'a',
    value: { 'stats/wins': 3, tag: null },
  });
  reject({ type: 'merge', writeId: W, path: 'a', value: 5 });
  reject({ type: 'merge', writeId: W, path: 'a', value: [1, 2] });
  assert.equal(reject({ type: 'merge', writeId: W, path: 'a', value: { '': 1 } }).code, 'BADPATH');
  assert.equal(reject({ type: 'merge', writeId: W, path: 'a', value: { 'x.y': 1 } }).code, 'BADPATH');
  assert.equal(reject({ type: 'merge', writeId: W, path: 'a', value: { 'x//y': 1 } }).code, 'BADPATH');
});

test('cas requires expectedRev', () => {
  assert.deepEqual(frame({ type: 'cas', writeId: W, path: 'a/score', expectedRev: 184224, value: 51 }), {
    type: 'cas',
    writeId: W,
    path: 'a/score',
    expectedRev: 184224,
    value: 51,
  });
  reject({ type: 'cas', writeId: W, path: 'a', value: 1 });
  reject({ type: 'cas', writeId: W, path: 'a', expectedRev: -1, value: 1 });
});

// ---- §5 liveness -----------------------------------------------------------------

test('ping/pong require a numeric t', () => {
  assert.deepEqual(frame({ type: 'pong', t: 1756280000000 }), { type: 'pong', t: 1756280000000 });
  reject({ type: 'ping' });
  reject({ type: 'ping', t: 'now' });
});
